keylime/keylime_agent.py

```diff
@@ -627,6 +637,20 @@ def main():
             raise RuntimeError('agent_uuid is configured to use dmidecode, '
                                'but it\'s is not found on the system.')
 
+    # initialize the tmpfs partition to store keys if it isn't already available
+    secdir = secure_mount.mount()
```

Comment:
I do not dislike this compromise here, but at this point I will fully go to this path that @THS-on pointed:

Copy https://github.com/utkarsh2102/python-keylime/blob/master/debian/python3-keylime-agent.var-lib-keylime-secure.mount into https://github.com/keylime/keylime/tree/master/services, and document it (maybe in the keylime.conf)

That way the package can easily downgrade the user changing the conf and enabling the mount point.

Also, you mean that the secure_mount.mount() can be in the same place and you can skip the last `chown`.

In any case, as commented, this is a nice compromise.

Reply:
> In any case, as commented, this is a nice compromise.

What does this mean now? Do you want me to keep the compromise or undo it and require that everyone use the systemd config file.

The thing is also the WORK_DIR ( secdir = $WORK_DIR/secure) can be set like this (config.py):

```
if not REQUIRE_ROOT:
    DEFAULT_WORK_DIR = os.path.abspath(".")
else:
    DEFAULT_WORK_DIR = '/var/lib/keylime'
WORK_DIR = os.getenv('KEYLIME_DIR', DEFAULT_WORK_DIR)
```

I am afraid that removing the compromise considering this me** wouldn't work for someone who set KEYLIME_DIR.

Reply:
> > In any case, as commented, this is a nice compromise.
> 
> What does this mean now? Do you want me to keep the compromise or undo it and require that everyone use the systemd config file.

Is your call. I personally prefer the systemd idea (but that is me). I am totally fine if remains as it is. I can add the mount service as a patch in the distribution, or rethink this later.

> 
> The thing is also the WORK_DIR ( secdir = $WORK_DIR/secure) can be set like this (config.py):
> 
> ```
> if not REQUIRE_ROOT:
>     DEFAULT_WORK_DIR = os.path.abspath(".")
> else:
>     DEFAULT_WORK_DIR = '/var/lib/keylime'
> WORK_DIR = os.getenv('KEYLIME_DIR', DEFAULT_WORK_DIR)
> ```

IMHO `REQUIRE_ROOT` is totally wrong. Seems to me a flag that is only used for testing.

The right flag is the one that you introduced: the user:group make `REQUIRE_ROOT` meaningless (except for old scripts, CI and the test suite)

Also `REQUIRE_ROOT` and `KEYLIME_DIR` are totally orthogonal, I can execute root  or non-root in `/var/lib/keylime`, and root or non-root in `what-ever-KEYLIME_DIR-says`.

Should be there later a different PR to address this.

> I am afraid that removing the compromise considering this me** wouldn't work for someone who set KEYLIME_DIR.

Not sure to understand this. If we are afraid of breaking the agent when KEYLIME_DIR is set to `/var/lib/keylime` and all that is inside is owned by `root:root`, and we now upgrade the agent and change the conf to use `keylime:tss`, and now the agent fails to run, that is exactly what should happen!

The migration process should be documented and executed by the package or by the admin, but not by the agent.

But if you mean that there is some incompatibility of setting `KEYLIME_DIR` and this PR, I do not see it.

With this PR it is now expected that the installation will set `KEYLIME_DIR` for whatever `user:group` is in the config.

Reply:
> But if you mean that there is some incompatibility of setting KEYLIME_DIR and this PR, I do not see it.

I do not see an incompatibility with setting KEYLIME_DIR, actually the opposite. If we were to rely on the systemd script for mounting we wouldn't be able to drop privileges since the secure directory may still have root ownership then  unless we chown() it as this PR does.

Reply:
I have the feeling that I am misunderstanding something. There are like several directories here. One in KEYLIME_DIR, that will be created under installation, with the correct user.  There is the secure directory that as today can be pre-created by the user, or if is missing, it will be created as a mount with tmpfs.

In this scenario, if the secure dir was not created, it will not be cleaned by the agent, but if was created by the agent as a tmpfs, will be umounted.

So the agent sometimes have responsibility over secure and sometimes don't, depending of some config vars and preconditions (like if it was already present). I agree that this is a bit of a mess now and should be fixed.

With systemd is like the same. It should be expected that systemd will mount the secure in, maybe tmpfs (or in some weird encrypted fs, we do not know) and under the correct user permissions (if not, the agent will fail to write the payload and will die). It will be already present before the agent is executed, and will be umounted later after the agent is gone.

Basically we do not make managing secure a responsibility of the agent.